### PR TITLE
Optimize `take_bits` / `take_boolean`: up to -25%

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #https://github.com/apache/arrow-rs/issues/6621

# Rationale for this change

Further optimization

```
take bool 1024          time:   [864.69 ns 865.71 ns 866.55 ns]
                        change: [-25.978% -25.819% -25.660%] (p = 0.00 < 0.05)
                        Performance has improved.

take bool null indices 1024
                        time:   [1.5549 µs 1.5560 µs 1.5572 µs]
                        change: [-16.412% -15.787% -15.182%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) low mild
  6 (6.00%) high mild
  7 (7.00%) high severe

take bool null values 1024
                        time:   [1.6879 µs 1.6897 µs 1.6916 µs]
                        change: [-11.554% -11.427% -11.294%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

take bool null values null indices 1024
                        time:   [3.5142 µs 3.5281 µs 3.5463 µs]
                        change: [-16.009% -15.478% -14.934%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) low mild
  8 (8.00%) high mild
  2 (2.00%) high severe
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-select/src/take.rs [hunk 1]
```diff
@@ -459,11 +461,10 @@ fn take_bytes<T: ByteArrayType, IndexType: ArrowPrimitiveType>(
 ) -> Result<GenericByteArray<T>, ArrowError> {
     let data_len = indices.len();
 
-    let bytes_offset = (data_len + 1) * std::mem::size_of::<T::Offset>();
-    let mut offsets = MutableBuffer::new(bytes_offset);
+    let mut offsets = Vec::with_capacity(data_len + 1);
     offsets.push(T::Offset::default());
 
-    let mut values = MutableBuffer::new(0);
+    let mut values = Vec::new();
```
Comment (tustvold) [inline arrow-select/src/take.rs:467]: This change seems unrelated to the benchmarks?

**Review by tustvold — APPROVED** [review]

File: arrow-select/src/take.rs (hunk 1 above)
Comment (Dandandan) [reply arrow-select/src/take.rs]: Removed this part - was looking at `take_bytes` as well but results are a bit mixed (might be benchmark related).

**Comment by Dandandan** [comment]
I found this also improves take for primitive / stringview with null values (15 - 20 %, added results in PR comment)
